README.md
# imagetobrailleart

##### A python library to convert image or text into braille art.

Convert any image or text into braille. I have used openCV, PIL and braillegraph by  [Chris bouchard] but i have slightly upgraded his library.

# Installation

This package is hosted on PyPI, so installation should be as simple as
```
$ pip install imagetobrailleart
```
Note that this package requires at least Python 3.3, so if your default Python installation is still Python 2, make sure you use pip3.

openCV and pillow are necessary for this library. They will be automatically installed in package. If it is not installed or not working then follow these commands.

```
$ pip install opencv-python==4.5.3.56
``` 

```
$ pip install pillow==8.3.1
``` 

Again, use python3 if necessary.

# Usage

You can either use ```imagetobraille``` or ```texttobraille```

Import imagetobrailleart to start using it.

### Image to Braille

```
imagetobraille("location_of_image", size = 100, inverse = 0)
```
100 is the minimum size we recommend and to inverse the image use 1 and if you don't want ot inverse it then use 0 or left it blank.

Example to print face of a girl:
```
>>> import imagetobrailleart as itb
>>> a = itb.imagetobraille("C:\\Users\\yash\\Desktop\\face2.jpg") 
>>> print(a)
```
![GIRL braille art](https://i.imgur.com/XYxyM7b.png)

### Text to Braille

```
texttobraille("font_location","text", size = 100, inverse = 0)
```
To inverse the image use 1 and if you don't want ot inverse it then use 0 or left it blank.

Example to print HEY:
```
>>> import imagetobrailleart as itb
>>> a = itb.texttoimage("arial.tff","HEY", 100, 0)
>>> print(a)
```
![HEY braille art](https://i.imgur.com/FmcX7lM.png)

## License

MIT

**Free Code, Hell Yeah!**

[//]: # (These are reference links used in the body of this note and get stripped out when the markdown processor does its job. There is no need to format nicely because it shouldn't be seen. Thanks SO - http://stackoverflow.com/questions/4823468/store-comments-in-markdown-syntax)

   [Chris bouchard]: <https://github.com/chrisbouchard/braillegraph>

CHANGELOG.txt
Change Log
==========

0.0.5 (13/12/2021)
-------------------


braillegraph.py
'''A library for creating graphs using Unicode braille characters

https://pypi.python.org/pypi/braillegraph

Someone on reddit posted a screenshot of their xmobar setup, which used braille
characters to show the loads of their four processor cores, as well as several
other metrics. I was impressed that you could fit so much data into a single
line. I immediately set out to implement braille bar graphs for myself.

The characters this script outputs are in the Unicode Braille Patterns section,
code points 0x2800 through 0x28FF. Not all fonts support these characters, so
if you can't see the examples below check your font settings.

There are two ways to use this package: imported in Python code, or as a
command line script.

To use the package in Python, import it and use the vertical_graph and
horizontal_graph functions.

    >>> from braillegraph import vertical_graph, horizontal_graph
    >>> vertical_graph([3, 1, 4, 1])
    '⡯⠥'
    >>> horizontal_graph([3, 1, 4, 1])
    '⣆⣇'

To use the package as a script, run it as

    % python -m braillegraph vertical 3 1 4 1 5 9 2 6
    ⡯⠥
    ⣿⣛⣓⠒⠂
    % python -m braillegraph horizontal 3 1 4 1 5 9 2 6
    ⠀⠀⢀
    ⠀⠀⣸⢠
    ⣆⣇⣿⣼

For a description of the arguments and flags, run

    % python -m braillegraph --help
'''

import itertools


# Only export this function. Everything else is an implementation detail.
__all__ = ['vertical_graph', 'horizontal_graph']


# The Unicode codepoint for an empty braille blocks (no raised dots)
_BRAILLE_EMPTY_BLOCK = 0x2800

# The offsets to add for dots. The dots are represented as an 8-bit bitfield as
# follows:
#
#     0x01  0x08
#     0x02  0x10
#     0x04  0x20
#     0x40  0x80
#
# Half rows only have a left-most dot. Full rows have both dots.
_BRAILLE_HALF_ROW = [0x01, 0x02, 0x04, 0x40]
_BRAILLE_HALF_ROW_RIGHT = [0x08, 0x10, 0x20, 0x80]
_BRAILLE_FULL_ROW = [0x09, 0x12, 0x24, 0xC0]

# Full columns have all four dots. Partial columns have dots starting from the
# bottom of the block. Note that _BRAILLE_PARTIAL_COL[x][y] represents (y + 1)
# dots.
_BRAILLE_FULL_COL = [0x47, 0xB8]
_BRAILLE_PARTIAL_COL = [
    [0x40, 0x44, 0x46],
    [0x80, 0xA0, 0xB0]
]


def _chunk(iterable, size):
    '''Split an iterable into chunks of a fixed size.'''
    # We're going to use some star magic to chunk the iterable. We create a
    # copy of the iterator size times, then pull a value from each to form a
    # chunk. The last chunk may have some trailing Nones if the length of the
    # iterable isn't a multiple of size, so we filter them out.
    #
    # pylint: disable=star-args
    args = (iter(iterable),) * size

    return (
        itertools.takewhile(lambda x: x is not None, group)
        for group in itertools.zip_longest(*args)
    )


def _matrix_add_column(matrix, column, default=0):
    '''Given a matrix as a list of lists, add a column to the right, filling in
    with a default value if necessary.
    '''
    extra_rows_needed = len(column) - len(matrix)
    width = len(matrix[0]) if matrix else 0
    offset = 0

    # If we need extra rows, add them to the top of the matrix.
    if extra_rows_needed > 0:
        for _ in range(extra_rows_needed):
            matrix.insert(0, [default] * width)

    # If the column is shorter, we'll need to shift it down.
    if extra_rows_needed < 0:
        offset = -extra_rows_needed
        column = ([default] * offset) + column

    for index, value in enumerate(column):
        matrix[index].append(value)


def vertical_graph(*args, sep='\n'):
    '''Consume an iterable of integers and produce a vertical bar graph using
    braille characters. If the iterable contains more than four integers, it
    will be chunked into groups of four, separated with newlines by default.

        >>> vertical_graph([1, 2, 3, 4])
        '⣷⣄'
        >>> vertical_graph([1, 2, 3, 4, 5, 6])
        '⣷⣄\\n⠛⠛⠓'
        >>> print(vertical_graph([1, 2, 3, 4, 5, 6]))
        ⣷⣄
        ⠛⠛⠓

    Alternately, the arguments can be passed directly:

        >>> vertical_graph(1, 2, 3, 4)
        '⣷⣄'

    The optional sep parameter controls how groups are separated. If sep is not
    passed (or if it is None), they are put on their own lines. For example, to
    keep everything on one line, space could be used:

        >>> vertical_graph(3, 1, 4, 1, 5, 9, 2, 6, sep=' ')
        '⡯⠥ ⣿⣛⣓⠒⠂'
    '''
    lines = []

    # If the arguments were passed as a single iterable, pull it out.
    # Otherwise, just use them as-is.
    if len(args) == 1:

        bars = args[0]
            
    else:
        bars = args

    # Make sure we use the default
    if sep is None:
        sep = '\n'

    # Break the bars into groups of four, one for each row in the braille
    # blocks.

    for bar_group in _chunk(bars, 4):
        line = []
        
        for braille_row, bar_value in enumerate(bar_group):
            # The number of full braille blocks needed to draw this bar. Each
            # block is two dots wide.
            # print(braille_row)
            if bar_value<0:
                bar_value = 1
                sign = -1
            else:
                sign = 1
            
            # print(f"{bar_value} {sign}")
                
            full_blocks_needed = bar_value // 2

            # The number of braille blocks needed to draw this bar. The second
            # term accounts for a possible half row.
            blocks_needed = full_blocks_needed + (bar_value % 2)

            # The number of braille blocks we'll need to append to the current
            # line to accomodate this bar
            extra_blocks_needed = blocks_needed - len(line)

            # If we need extra blocks, add them.
            if extra_blocks_needed > 0:
                line.extend([_BRAILLE_EMPTY_BLOCK] * extra_blocks_needed)

            # Fill in the majority of the bar with full braille rows (two dots).
            for block_index in range(full_blocks_needed):
                line[block_index] += _BRAILLE_FULL_ROW[braille_row]

            # If the bar's value is odd, we'll need to add a single dot at the
            # end.
            if bar_value % 2:
                if sign == -1:
                    line[full_blocks_needed] += _BRAILLE_HALF_ROW_RIGHT[braille_row]
                else:
                    line[full_blocks_needed] += _BRAILLE_HALF_ROW[braille_row]
                    

        # Wrap up this line by converting all the code points to characters
        # and concatenating them.
        lines.append(''.join(chr(code) for code in line))

    # Join all the lines to make the final graph
    return sep.join(lines)


def horizontal_graph(*args):
    '''Consume an iterable of integers and produce a horizontal bar graph using
    braille characters.

        >>> horizontal_graph([1, 2, 3, 4])
        '⣠⣾'
        >>> horizontal_graph([1, 2, 3, 4, 5, 6])
        '⠀⠀⣠\\n⣠⣾⣿'
        >>> print(horizontal_graph([1, 2, 3, 4, 5, 6]))
        ⠀⠀⣠
        ⣠⣾⣿

    Alternately, the arguments can be passed directly:

        >>> horizontal_graph(1, 2, 3, 4)
        '⣠⣾'
    '''
    lines = []

    # If the arguments were passed as a single iterable, pull it out.
    # Otherwise, just use them as-is.
    if len(args) == 1:

        bars = args[0]
    else:
        bars = args

    # Break the bars into groups of two, one for each column in the braille
    # blocks.
    for bar_group in _chunk(bars, 2):
        column = []

        for braille_col, bar_value in enumerate(bar_group):
            # The number of full braille blocks needed to draw this bar. Each
            # block is four dots tall.
            full_blocks_needed = bar_value // 4

            # The number of braille blocks needed to draw this bar. This
            # accounts for a possible partial block.
            blocks_needed = full_blocks_needed + (1 if bar_value % 4 else 0)

            # The number of new lines we'll need to prepend to accomodate this
            # bar
            extra_blocks_needed = blocks_needed - len(column)

            # If we need extra blocks, add them.
            column = ([_BRAILLE_EMPTY_BLOCK] * extra_blocks_needed) + column

            # Fill in the majority of the column with full braille colums (four
            # dots).
            for block_index in range(-full_blocks_needed, 0, 1):
                column[block_index] += _BRAILLE_FULL_COL[braille_col]

            # If we need a partial column, fill it in.
            if bar_value % 4:
                partial_index = (bar_value % 4) - 1
                column[-blocks_needed] += _BRAILLE_PARTIAL_COL[braille_col][partial_index]

        # Add this column to the lines.
        _matrix_add_column(lines, column, default=_BRAILLE_EMPTY_BLOCK)

    # Convert all the code points into characters, concatenate them into lines,
    # then concatenate all the lines to make the final graph.
    return '\n'.join(''.join(chr(code) for code in line) for line in lines)

imagetobrailleart.py
# Python script to convert the given image
# into a dotted text using opencv
  
# import the required modules
import cv2
from PIL import Image, ImageDraw, ImageFont
import os
import requests

from modifiedbraillegraph import vertical_graph, horizontal_graph

# Read the image
def imagetobraille(image_location = "", size =100, inverse = 0):
    
    if image_location == "":
        return "Image location not specified"

    img = cv2.imread(image_location, 0)

    # Apply median blur
    try:
        img = cv2.medianBlur(img, 5)
    except:
        return "Image location incorrect"
    
    # Apply MEAN thresholding to get refined edges
    image = cv2.adaptiveThreshold(
        img, 255, cv2.ADAPTIVE_THRESH_MEAN_C, cv2.THRESH_BINARY, 11, 2)
    
    # Preserve the ratio
    ratio = len(image)/len(image[0])
    # Assign new width and calculate new height
    new_height = size
    new_width = int(ratio*new_height)
    # Resize the image
    image = cv2.resize(image, (new_height, new_width))

    fillone = 0
    filltwo = 2

    if inverse == 1:
        fillone = 2
        filltwo = 0
    # Iterate over the array and print the dark pixels
    # or we can use any other symbol too.
    outputart = ""
    b=0

    pixellist1 = []
    pixellist2 = []
    pixellist3 = []
    pixellist4 = []

    noofdotslist1 = []
    noofdotslist2 = []
    noofdotslist3 = []
    noofdotslist4 = []

    # print(image[0])
    for i in range(len(image)):

        for j in range(len(image[0])):
            if b==0:
                if image[i, j] < 200:
                    pixellist1.append(0)
                else:
                    pixellist1.append(1)
            elif b==1:
                if image[i, j] < 200:
                    pixellist2.append(0)
                else:
                    pixellist2.append(1)
            elif b==2:
                if image[i, j] < 200:
                    pixellist3.append(0)
                else:
                    pixellist3.append(1)
            elif b==3:
                if image[i, j] < 200:
                    pixellist4.append(0)
                else:
                    pixellist4.append(1)

        if len(pixellist1)%2 !=0:
            pixellist1.append(1)
            pixellist2.append(1)
            pixellist3.append(1)
            pixellist4.append(1)

        b += 1
        
        if b==4:
            for k in range(0,len(pixellist1),2):
                
                if pixellist1[k] == 1 and pixellist1[k+1] == 1:
                    noofdotslist1.append(fillone)
                elif pixellist1[k] == 1 and pixellist1[k+1] == 0:
                    noofdotslist1.append(-1)
                elif pixellist1[k] == 0 and pixellist1[k+1] == 1:
                    noofdotslist1.append(1)
                else:
                    noofdotslist1.append(filltwo)

                if pixellist2[k] == 1 and pixellist2[k+1] == 1:
                    noofdotslist2.append(fillone)
                elif pixellist2[k] == 1 and pixellist2[k+1] == 0:
                    noofdotslist2.append(-1)
                elif pixellist2[k] == 0 and pixellist2[k+1] == 1:
                    noofdotslist2.append(1)
                else:
                    noofdotslist2.append(filltwo)

                if pixellist3[k] == 1 and pixellist3[k+1] == 1:
                    noofdotslist3.append(fillone)
                elif pixellist3[k] == 1 and pixellist3[k+1] == 0:
                    noofdotslist3.append(-1)
                elif pixellist3[k] == 0 and pixellist3[k+1] == 1:
                    noofdotslist3.append(1)
                else:
                    noofdotslist3.append(filltwo)

                if pixellist4[k] == 1 and pixellist4[k+1] == 1:
                    noofdotslist4.append(fillone)
                elif pixellist4[k] == 1 and pixellist4[k+1] == 0:
                    noofdotslist4.append(-1)
                elif pixellist4[k] == 0 and pixellist4[k+1] == 1:
                    noofdotslist4.append(1)
                else:
                    noofdotslist4.append(filltwo)

            try:
                for l in range(0,len(noofdotslist1)):
                    
                    if noofdotslist1[l] or noofdotslist2[l] or noofdotslist3[l] or noofdotslist4[l] != 0:
                        outputart += vertical_graph([noofdotslist1[l], noofdotslist2[l], noofdotslist3[l], noofdotslist4[l]])
                    else:    
                        outputart +=" "
            except:
                print("KYA HUA")

            outputart += "\n"  

            pixellist1.clear()
            pixellist2.clear()
            pixellist3.clear()
            pixellist4.clear()

            noofdotslist1.clear()
            noofdotslist2.clear()
            noofdotslist3.clear()
            noofdotslist4.clear()
            
            b=0

    # cv2.waitKey(0)
    return outputart

#download font file in .tff format
def texttobraille(font_location : str,text : str, size = 100, inverse = 0):
    image_url = "https://i.imgur.com/CykXjV9.jpg"    
    im = Image.open(requests.get(image_url, stream=True).raw)
    im.save("imagetobrailleartblankpic.jpg")

    textimg = Image.open('imagetobrailleartblankpic.jpg')
    d1 = ImageDraw.Draw(textimg)
    myFont = ImageFont.truetype(font_location, 100)
    d1.text((0, 0), text, fill =(0, 0, 0),font = myFont)

    textimg.save("{}.jpg".format(text))
    brailleart = imagetobraille(text+".jpg", size, inverse) 
    textimg.show()   
    os.remove("{}.jpg".format(text))
    os.remove("imagetobrailleartblankpic.jpg")
    return brailleart

imagetodottedtext.py
# Python script to convert the given image
# into a dotted text using opencv
  
# import the required modules
import cv2
from braillegraph import vertical_graph, horizontal_graph
from PIL import Image, ImageDraw, ImageFont
import os

# Read the image
def brailletoart(image_name = "face2", size =100, inverse = "n", pngorjpg = "j" ):
    
    if pngorjpg == "j":
            image_name = "images\\"+ image_name + ".jpg"
    elif pngorjpg == "p":
            image_name = "images\\"+ image_name + ".png"         

    img = cv2.imread(image_name, 0)

    # Apply median blur
    img = cv2.medianBlur(img, 5)
    
    # Apply MEAN thresholding to get refined edges
    image = cv2.adaptiveThreshold(
        img, 255, cv2.ADAPTIVE_THRESH_MEAN_C, cv2.THRESH_BINARY, 11, 2)

    # cv2.imshow("image", image)
    
    # Preserve the ratio
    ratio = len(image)/len(image[0])
    # Assign new width and calculate new height
    new_height = size
    new_width = int(ratio*new_height)
    # print(f"{new_width} {new_height}")
    # Resize the image
    image = cv2.resize(image, (new_height, new_width))

    fillone = 0
    filltwo = 2

    if inverse == "y":
        fillone = 2
        filltwo = 0
    # Iterate over the array and print the dark pixels
    # or we can use any other symbol too.
    a = ""
    b=0

    pixellist1 = []
    pixellist2 = []
    pixellist3 = []
    pixellist4 = []

    noofdotslist1 = []
    noofdotslist2 = []
    noofdotslist3 = []
    noofdotslist4 = []

    # print(image[0])
    for i in range(len(image)):

        for j in range(len(image[0])):
            # a += "   " if image[i, j] < 200 else ".  "
            if b==0:
                if image[i, j] < 200:
                    pixellist1.append(0)
                else:
                    pixellist1.append(1)
            elif b==1:
                if image[i, j] < 200:
                    pixellist2.append(0)
                else:
                    pixellist2.append(1)
            elif b==2:
                if image[i, j] < 200:
                    pixellist3.append(0)
                else:
                    pixellist3.append(1)
            elif b==3:
                if image[i, j] < 200:
                    pixellist4.append(0)
                else:
                    pixellist4.append(1)

        if len(pixellist1)%2 !=0:
            pixellist1.append(1)
            pixellist2.append(1)
            pixellist3.append(1)
            pixellist4.append(1)

        b += 1
        
        if b==4:
            for k in range(0,len(pixellist1),2):
                
                if pixellist1[k] == 1 and pixellist1[k+1] == 1:
                    noofdotslist1.append(fillone)
                elif pixellist1[k] == 1 and pixellist1[k+1] == 0:
                    noofdotslist1.append(-1)
                elif pixellist1[k] == 0 and pixellist1[k+1] == 1:
                    noofdotslist1.append(1)
                else:
                    noofdotslist1.append(filltwo)

                if pixellist2[k] == 1 and pixellist2[k+1] == 1:
                    noofdotslist2.append(fillone)
                elif pixellist2[k] == 1 and pixellist2[k+1] == 0:
                    noofdotslist2.append(-1)
                elif pixellist2[k] == 0 and pixellist2[k+1] == 1:
                    noofdotslist2.append(1)
                else:
                    noofdotslist2.append(filltwo)

                if pixellist3[k] == 1 and pixellist3[k+1] == 1:
                    noofdotslist3.append(fillone)
                elif pixellist3[k] == 1 and pixellist3[k+1] == 0:
                    noofdotslist3.append(-1)
                elif pixellist3[k] == 0 and pixellist3[k+1] == 1:
                    noofdotslist3.append(1)
                else:
                    noofdotslist3.append(filltwo)

                if pixellist4[k] == 1 and pixellist4[k+1] == 1:
                    noofdotslist4.append(fillone)
                elif pixellist4[k] == 1 and pixellist4[k+1] == 0:
                    noofdotslist4.append(-1)
                elif pixellist4[k] == 0 and pixellist4[k+1] == 1:
                    noofdotslist4.append(1)
                else:
                    noofdotslist4.append(filltwo)

            try:
                for l in range(0,len(noofdotslist1)):
                    
                    if noofdotslist1[l] or noofdotslist2[l] or noofdotslist3[l] or noofdotslist4[l] != 0:
                        a += vertical_graph([noofdotslist1[l], noofdotslist2[l], noofdotslist3[l], noofdotslist4[l]])
                    else:    
                        a +=" "
            except:
                print("KYA HUA")

            a += "\n"  

            pixellist1.clear()
            pixellist2.clear()
            pixellist3.clear()
            pixellist4.clear()

            noofdotslist1.clear()
            noofdotslist2.clear()
            noofdotslist3.clear()
            noofdotslist4.clear()
            
            b=0


    print("\n"*5)
    print(a)
    # outputimg
    print("Done")
    cv2.waitKey(0)

def textedimage(text : str):
    textimg = Image.open('images/back.jpg')
    d1 = ImageDraw.Draw(textimg)
    myFont = ImageFont.truetype("font.ttf", 100)
    d1.text((5, 5), text.upper(), fill =(0, 0, 0), font= myFont)

    textimg.save("images\\{}.jpg".format(text))


while True:
    try:
        command = input("Enter command (t if text none if not text, image/text name, size, y for inverse and y for non inverse, j for jpg and p for png,): ")
        commandlist = command.split()

        if commandlist[0] == "t":
            
            texttoprint = ""
            for text in range(1,len(commandlist)-2):
                texttoprint = texttoprint + "{} ".format(commandlist[text])
            
            texttoprint = texttoprint[:-1]

            textedimage(texttoprint)

            brailletoart(texttoprint, int(commandlist[-2]), commandlist[-1])    
            os.remove("images\\{}.jpg".format(texttoprint))
        else:
            brailletoart(commandlist[0], commandlist[1], int(commandlist[2]), commandlist[3])    

    except BaseException as error:
        print('An exception occurred: {}'.format(error))
        continue
